core_lib/src/structure/layerbitmap.cpp

```diff
@@ -202,24 +202,41 @@ void LayerBitmap::loadDomElement(const QDomElement& element, QString dataDirPath
     while (!imageTag.isNull())
     {
         QDomElement imageElement = imageTag.toElement();
-        if (!imageElement.isNull())
+        if (!imageElement.isNull() && imageElement.tagName() == "image")
         {
-            if (imageElement.tagName() == "image")
+            bool shouldLoad = true;
+
+            // Make sure src path is relative
+            shouldLoad &= QFileInfo(imageElement.attribute("src")).isRelative();
+            QFileInfo fi(dataDirPath, imageElement.attribute("src"));
+            // Make sure file is in data directory after resolving relative components and symlinks
+            QString canonicalPath = fi.canonicalFilePath();
+            fi.setFile(!canonicalPath.isEmpty() ? canonicalPath : fi.absoluteFilePath());
+            QDir dataDir(dataDirPath);
+            QDir ancestor = fi.dir();
+            while (ancestor != dataDir)
```

Comment:
Couldn't we move this to a method in Layer, so we don't have to have it duplicated three times?

Reply:
Yes very good point. I have refactored this into a separate function. It is in util because this is not exclusively used by layers now.